src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.OData.Edm;
+
+namespace Microsoft.AspNet.OData
+{
+    /// <summary>
+    /// Represents an <see cref="IEdmObject"/> that is a collection of <see cref="IEdmChangedObject"/>s.
+    /// </summary>
+    [NonValidatingParameterBinding]
+    public class EdmChangedObjectCollection<TStructuralType> : EdmChangedObjectCollection, ICollection<IEdmChangedObject<TStructuralType>>, IEdmObject
```

Comment:
>EdmChangedObjectCollection [](start = 63, length = 26)

Important to note that there are really two Collection<T> implementations here – our private _items Collection<IEdmChangedObject<TStructuralType>>and the Collection<IEdmChangedObject> that the EdmChangedObjectCollection derives from.  Each of these collections (currently) have their own storage.

If I call Add() on the derived EdmChangedObjectCollection<TStructuralType> it will call our implementation (which will add/remove it o/from our _items collection), but if I cast to EdmChangedObjectCollection and call Add() it will add to the contained collection, right?

To make sure that all methods add/remove/change values in our _items collection, we probably need to override in this class the protected virtuals implemented by the base Collection<IEdmChangedObject>:
        protected virtual void ClearItems();
        protected virtual void InsertItem(int index, T item);
        protected virtual void RemoveItem(int index);
        protected virtual void SetItem(int index, T item);
In our override, we should:
   1) Make sure the type is IEdmChangeObejct<TStructuralType>
   2) Implement over our internal Collection<IEdmChangedObject<TStructuralType>>  _items, rather than clearing/adding/removing/setting values on the Collection<IEdmChangedObject> that EdmChangedObjectCollection derives from.

Reply:
Does not appear that this has been addressed.  Looks like there are still two different underlying storage collections -- our private _items and the base Collection<T> underlying the base EdmChangedObjectCollection. The protected virtuals described above will guarantee that changes made to the base EdmChangedObjectCollection<T> are reflected in the derived class.

---
In reply to: [518419162](https://github.com/OData/WebApi/pull/2275#discussion_r518419162) [](ancestors = 518419162)